airtime_mvc/application/models/Block.php

```diff
@@ -1355,14 +1375,30 @@ public function getListOfFilesUnderLimit()
         $sizeOfInsert = count($insertList);
         
         // if block is not full and repeat_track is check, fill up more
+        // additionally still don't overflow the limit
         while (!$isBlockFull && $repeat == 1 && $sizeOfInsert > 0) {
         	Logging::debug("adding repeated tracks.");
         	Logging::debug("total time = " . $totalTime);
         	
             $randomEleKey = array_rand(array_slice($insertList, 0, $sizeOfInsert));
-            $insertList[] = $insertList[$randomEleKey];
-            $totalTime += $insertList[$randomEleKey]['length'];
-            $totalItems++;
+            // this will also allow the overflow of tracks so that time limited smart blocks will schedule until they
+            // are longer than the time limit rather than never scheduling past the time limit
+            if ($overflow == 1) {
+                $insertList[] = $insertList[$randomEleKey];
+                $totalTime += $insertList[$randomEleKey]['length'];
+                $totalItems++;
+            }
+            else {
+                $projectedTime = $totalTime + $insertList[$randomEleKey]['length'];
+                if ($projectedTime > $limit['time']) {
+                    $totalItems++;
+                }
+                else {
+                    $insertList[] = $insertList[$randomEleKey];
```

Comment:
We might want to refactor the `getListOfFilesUnderLimit()` method at some point since it's becoming unwieldy (ie. the contents of this else are an exact copy of lines 1378-1389 above. I'm fine with not refactoring it in the PR though.